**Data Preprosseing**

In [1]:
import numpy as np
import pandas as pd


In [2]:
data = pd.read_csv('laptop_price.csv',encoding='latin-1')

In [3]:
data.head()

laptop_ID Company      Product   TypeName  Inches  \
0          1   Apple  MacBook Pro  Ultrabook    13.3   
1          2   Apple  Macbook Air  Ultrabook    13.3   
2          3      HP       250 G6   Notebook    15.6   
3          4   Apple  MacBook Pro  Ultrabook    15.4   
4          5   Apple  MacBook Pro  Ultrabook    13.3   

                     ScreenResolution                         Cpu   Ram  \
0  IPS Panel Retina Display 2560x1600        Intel Core i5 2.3GHz   8GB   
1                            1440x900        Intel Core i5 1.8GHz   8GB   
2                   Full HD 1920x1080  Intel Core i5 7200U 2.5GHz   8GB   
3  IPS Panel Retina Display 2880x1800        Intel Core i7 2.7GHz  16GB   
4  IPS Panel Retina Display 2560x1600        Intel Core i5 3.1GHz   8GB   

                            Gpu  OpSys  Weight  Price_euros  
0  Intel Iris Plus Graphics 640  macOS  1.37kg      1339.69  
1        Intel HD Graphics 6000  macOS  1.34kg       898.94  
2         Intel HD Graphics 620  No OS  1.86kg       575.00  
3            AMD Radeon Pro 455  macOS  1.83kg      2537.45  
4  Intel Iris Plus Graphics 650  macOS  1.37kg      1803.60

In [4]:
data.shape


(1303, 12)

In [5]:
data.isnull().sum()

laptop_ID           0
Company             0
Product             0
TypeName            0
Inches              0
ScreenResolution    0
Cpu                 0
Ram                 0
Gpu                 0
OpSys               0
Weight              0
Price_euros         0
dtype: int64

In [25]:
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1303 entries, 0 to 1302
Data columns (total 12 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   Company           1303 non-null   object 
 1   TypeName          1303 non-null   object 
 2   Inches            1303 non-null   float64
 3   ScreenResolution  1303 non-null   object 
 4   Cpu               1303 non-null   object 
 5   Ram               1303 non-null   int32  
 6   Gpu               1303 non-null   object 
 7   OpSys             1303 non-null   object 
 8   Weight            1303 non-null   float32
 9   Price_euros       1303 non-null   float64
 10  TouchScreen       1303 non-null   int64  
 11  IPS               1303 non-null   int64  
dtypes: float32(1), float64(2), int32(1), int64(2), object(6)
memory usage: 112.1+ KB


First we need to check are there any column we can simply remove(no effect for the price) > here we have id coloumn 

In [ ]:
data = data.drop('laptop_ID',axis='columns')


In [12]:
data.head()

Company      Product   TypeName  Inches                    ScreenResolution  \
0   Apple  MacBook Pro  Ultrabook    13.3  IPS Panel Retina Display 2560x1600   
1   Apple  Macbook Air  Ultrabook    13.3                            1440x900   
2      HP       250 G6   Notebook    15.6                   Full HD 1920x1080   
3   Apple  MacBook Pro  Ultrabook    15.4  IPS Panel Retina Display 2880x1800   
4   Apple  MacBook Pro  Ultrabook    13.3  IPS Panel Retina Display 2560x1600   

                          Cpu   Ram                           Gpu  OpSys  \
0        Intel Core i5 2.3GHz   8GB  Intel Iris Plus Graphics 640  macOS   
1        Intel Core i5 1.8GHz   8GB        Intel HD Graphics 6000  macOS   
2  Intel Core i5 7200U 2.5GHz   8GB         Intel HD Graphics 620  No OS   
3        Intel Core i7 2.7GHz  16GB            AMD Radeon Pro 455  macOS   
4        Intel Core i5 3.1GHz   8GB  Intel Iris Plus Graphics 650  macOS   

   Weight  Price_euros  
0  1.37kg      1339.69  
1  1.34kg       898.94  
2  1.86kg       575.00  
3  1.83kg      2537.45  
4  1.37kg      1803.60

Now we need to check the object columns and convert those into numerical types

Here we have ram and weight column , we can simply remove kg and GB parts from those and convert to numerical types

In [ ]:
data['Ram'] = data['Ram'].str.replace('GB','').astype('int32')
data['Weight'] = data['Weight'].str.replace('kg','').astype('float32')
data.info()

Now we should check other object columns one by one , and identify different types of each column have , if there are more types we have to reduce the types 

In [18]:
data['Company'].value_counts()


Company
Dell       297
Lenovo     297
HP         274
Asus       158
Acer       103
MSI         54
Other       51
Toshiba     48
Apple       21
Name: count, dtype: int64

here we have so many types if we do the one hot encoding there are 19 columns , so we get the all types together having value counts lower than 10

In [17]:
def company_categorize(company):
    if data['Company'].value_counts()[company] < 10:
        return 'Other'
    else:
        return company

data['Company'] = data['Company'].apply(company_categorize)

Now we check product column

In [20]:
data['Product'].value_counts()
len(data['Product'].value_counts())

618

here there are 618 types so so product should differ almost over every column so , so it does not effect price more so we can simply remove the column

In [21]:
data=data.drop('Product',axis='columns')

now we check TypeName column 

In [22]:
data['TypeName'].value_counts()

TypeName
Notebook              727
Gaming                205
Ultrabook             196
2 in 1 Convertible    121
Workstation            29
Netbook                25
Name: count, dtype: int64

here it has only six so for one hot encoding it does not matter so we can keep it without changing

Inches is a numerical column , so we move to resolution column

In [23]:
data['ScreenResolution'].value_counts()

ScreenResolution
Full HD 1920x1080                                507
1366x768                                         281
IPS Panel Full HD 1920x1080                      230
IPS Panel Full HD / Touchscreen 1920x1080         53
Full HD / Touchscreen 1920x1080                   47
1600x900                                          23
Touchscreen 1366x768                              16
Quad HD+ / Touchscreen 3200x1800                  15
IPS Panel 4K Ultra HD 3840x2160                   12
IPS Panel 4K Ultra HD / Touchscreen 3840x2160     11
4K Ultra HD / Touchscreen 3840x2160               10
IPS Panel 1366x768                                 7
Touchscreen 2560x1440                              7
4K Ultra HD 3840x2160                              7
IPS Panel Retina Display 2304x1440                 6
IPS Panel Retina Display 2560x1600                 6
Touchscreen 2256x1504                              6
IPS Panel Quad HD+ / Touchscreen 3200x1800         6
IPS Panel Touchscreen 2560x14

here there are lot of types , so I check weather it has ips pannel and touch screen or not as a new column and remove this Screen resolution column 

In [24]:
data['TouchScreen']=data['ScreenResolution'].apply(lambda x:1 if 'Touchscreen' in x else 0)

data['IPS'] = data['ScreenResolution'].apply(lambda x:1 if 'IPS' in x else 0)

In [26]:
data=data.drop('ScreenResolution',axis='columns')

Cpu cloumn

In [52]:
data['Cpu'].value_counts()

Cpu
Intel Core i5 7200U 2.5GHz       190
Intel Core i7 7700HQ 2.8GHz      146
Intel Core i7 7500U 2.7GHz       134
Intel Core i7 8550U 1.8GHz        73
Intel Core i5 8250U 1.6GHz        72
                                ... 
Intel Core i5 7200U 2.70GHz        1
Intel Core M M7-6Y75 1.2GHz        1
Intel Core M 6Y54 1.1GHz           1
AMD E-Series 9000 2.2GHz           1
Samsung Cortex A72&A53 2.0GHz      1
Name: count, Length: 118, dtype: int64

now we get first 3 words to initalize the cpu type

In [53]:
data['Cpu_type']=data['Cpu'].apply(lambda x:' '.join(x.split()[0:3]))
data['Cpu_type'].value_counts()

Cpu_type
Intel Core i7               527
Intel Core i5               423
Intel Core i3               136
Intel Celeron Dual           80
Intel Pentium Quad           27
Intel Core M                 19
AMD A9-Series 9420           12
AMD A6-Series 9220            8
Intel Celeron Quad            8
AMD A12-Series 9720P          7
Intel Atom x5-Z8350           5
AMD A8-Series 7410            4
Intel Atom x5-Z8550           4
AMD A9-Series 9410            3
Intel Pentium Dual            3
AMD Ryzen 1700                3
AMD A9-Series A9-9420         2
AMD E-Series E2-9000e         2
AMD A10-Series A10-9620P      2
AMD A6-Series A6-9220         2
AMD E-Series 7110             2
AMD A10-Series 9620P          2
AMD A10-Series 9600P          2
Intel Xeon E3-1505M           2
Intel Xeon E3-1535M           2
Intel Atom X5-Z8350           2
Intel Atom x5-Z8300           1
AMD E-Series 6110             1
AMD E-Series 9000e            1
AMD E-Series E2-6110          1
AMD FX 9830P                  1

now we get types intel core i3 i5 i7 and amd and other to reduce types

In [54]:
def cpu_categorize(cpu):
    if cpu == 'Intel Core i7' or cpu == 'Intel Core i5' or cpu == 'Intel Core i3':
        return cpu
    else:
        if cpu.split()[0]=='AMD':
            return 'AMD'
        else:
            return 'Other'

data['Cpu_type']=data['Cpu_type'].apply(cpu_categorize)

In [55]:
data['Cpu_type'].value_counts()

Cpu_type
Intel Core i7    527
Intel Core i5    423
Other            155
Intel Core i3    136
AMD               62
Name: count, dtype: int64

In [57]:
data=data.drop('Cpu',axis='columns') 

Now we move to the gpu column 

In [58]:
data['Gpu'].value_counts()

Gpu
Intel HD Graphics 620      281
Intel HD Graphics 520      185
Intel UHD Graphics 620      68
Nvidia GeForce GTX 1050     66
Nvidia GeForce GTX 1060     48
                          ... 
Nvidia Quadro M500M          1
AMD Radeon R7 M360           1
Nvidia Quadro M3000M         1
Nvidia GeForce 960M          1
ARM Mali T860 MP4            1
Name: count, Length: 110, dtype: int64

We only categorize Intel , AMD , And Nvidea

In [59]:
data['Gpu_brand']=data['Gpu'].apply(lambda x:x.split()[0])

In [60]:
data['Gpu_brand'].value_counts()

Gpu_brand
Intel     722
Nvidia    400
AMD       180
ARM         1
Name: count, dtype: int64

In [62]:
#here we have other type ARM we simply remove it
data=data[data['Gpu_brand']!='ARM']

In [63]:
 data=data.drop('Gpu',axis='columns')

In [65]:
data['OpSys'].value_counts()

OpSys
Windows 10      1072
No OS             66
Linux             62
Windows 7         45
Chrome OS         26
macOS             13
Mac OS X           8
Windows 10 S       8
Android            2
Name: count, dtype: int64

In [ ]:
data['Opsys_type']=data['OpSys'].apply(lambda x:'Windows' if 'Windows' in x else ('Mac' if 'Mac' or 'mac' in x else 'Other'))

In [68]:
data['Opsys_type'].value_counts()

Opsys_type
Windows    1125
Other       169
Mac           8
Name: count, dtype: int64

In [70]:
data=data.drop('OpSys',axis='columns')

In [71]:
data.head()

Company   TypeName  Inches  Ram  Weight  Price_euros  TouchScreen  IPS  \
0   Apple  Ultrabook    13.3    8    1.37      1339.69            0    1   
1   Apple  Ultrabook    13.3    8    1.34       898.94            0    0   
2      HP   Notebook    15.6    8    1.86       575.00            0    0   
3   Apple  Ultrabook    15.4   16    1.83      2537.45            0    1   
4   Apple  Ultrabook    13.3    8    1.37      1803.60            0    1   

        Cpu_type Gpu_brand Opsys_type  
0  Intel Core i5     Intel      Other  
1  Intel Core i5     Intel      Other  
2  Intel Core i5     Intel      Other  
3  Intel Core i7       AMD      Other  
4  Intel Core i5     Intel      Other